File: frame/v2.0/top.py

from build123d import *
from ocp_vscode import show

unit_size           = 8.0 # lego unit size
unit_height         = 9.6 # lego brick height
hole_inner_diameter = 4.8 # lego hole inner diameter
hole_outer_diameter = 6.2 # lego hole outer diameter

with BuildPart() as part:
    # База
    length = unit_size * 26
    width  = unit_size * 10
    height = unit_height * 2
    Box(length, width, height)
    
    # Вырезы
    x_pos = unit_size * -9
    y_pos = 0
    z_pos = 0
    with BuildPart(Location((x_pos, y_pos, z_pos)), mode=Mode.SUBTRACT):
        length = unit_size * 8
        width  = unit_size * 6
        height = unit_height * 2
        align  = (Align.CENTER, Align.CENTER, Align.CENTER)
        Box(length, width, height)
    
    x_pos = unit_size * 9
    y_pos = 0
    z_pos = 0
    with BuildPart(Location((x_pos, y_pos, z_pos)), mode=Mode.SUBTRACT):
        length = unit_size * 8
        width  = unit_size * 6
        height = unit_height * 2
        align  = (Align.CENTER, Align.CENTER, Align.CENTER)
        Box(length, width, height)
    
    x_pos = 0
    y_pos = 0
    z_pos = 0
    with BuildPart(Location((x_pos, y_pos, z_pos)), mode=Mode.SUBTRACT):
        length = 44
        width  = 22
        height = unit_height * 2
        align  = (Align.CENTER, Align.CENTER, Align.CENTER)
        Box(length, width, height)
    
    x_pos = 0
    y_pos = unit_size * -2.5
    z_pos = 0
    with BuildPart(Location((x_pos, y_pos, z_pos)), mode=Mode.SUBTRACT):
        length = unit_size * 26
        width  = unit_size * 5
        height = unit_height * 2
        align  = (Align.CENTER, Align.CENTER, Align.CENTER)
        Box(length, width, height)
    
    x_pos = unit_size * -7.5
    y_pos = 0
    z_pos = unit_height
    with BuildPart(Location((x_pos, y_pos, z_pos)), mode=Mode.SUBTRACT):
        length = unit_size * 7
        width  = unit_size * 10
        height = unit_height * 2
        align  = (Align.CENTER, Align.CENTER, Align.CENTER)
        Box(length, width, height)
    
    x_pos = unit_size * 7.5
    y_pos = 0
    z_pos = unit_height
    with BuildPart(Location((x_pos, y_pos, z_pos)), mode=Mode.SUBTRACT):
        length = unit_size * 7
        width  = unit_size * 10
        height = unit_height * 2
        align  = (Align.CENTER, Align.CENTER, Align.CENTER)
        Box(length, width, height)
    
    x_pos = 0
    y_pos = unit_size * 4
    z_pos = unit_height
    with BuildPart(Location((x_pos, y_pos, z_pos)), mode=Mode.SUBTRACT):
        length = unit_size * 8
        width  = unit_size * 2
        height = unit_height * 2
        align  = (Align.CENTER, Align.CENTER, Align.CENTER)
        Box(length, width, height)
    
    # Отверстия под крепежные винты
    points = [
        (unit_size * -12, unit_size * 4,  0),
        (unit_size * -12, unit_size * -4, 0),
        (unit_size * 12,  unit_size * 4,  0),
        (unit_size * 12,  unit_size * -4, 0),
    ]
    for point in points:
        with BuildPart(Location(point), mode=Mode.SUBTRACT):
            radius = 1.5
            height = unit_height * 2
            align  = (Align.CENTER, Align.CENTER, Align.CENTER)
            Cylinder(radius, height, align=align)
    
    # Утоплялки под крепежные винты
    z_pos = unit_height / 2
    points = [
        (unit_size * -12, unit_size * 4,  z_pos),
        (unit_size * -12, unit_size * -4, z_pos),
        (unit_size * 12,  unit_size * 4,  z_pos),
        (unit_size * 12,  unit_size * -4, z_pos),
    ]
    for point in points:
        with BuildPart(Location(point), mode=Mode.SUBTRACT):
            radius = 3
            height = unit_height
            align  = (Align.CENTER, Align.CENTER, Align.CENTER)
            Cylinder(radius, height, align=align)

    roundable_edges = part.edges().filter_by(Axis.Z)
    fillet(roundable_edges, radius=2)

filename = f'{__file__[:-3]}'

show(part.part, names=[filename])
export_stl(part.part, f'{filename}.stl')

File: frame/v2.0/replicad/top.js
const { makeBox, makeCylinder } = replicad;

const unitSize = 8.0;           // lego unit size
const unitHeight = 9.6;         // lego brick height
const holeInnerDiameter = 4.8;  // lego hole inner diameter
const holeOuterDiameter = 6.2;  // lego hole outer diameter

const main = () => {
    // Base
    const length = unitSize * 26;
    const width = unitSize * 10;
    const height = unitHeight * 2;
    
    let part = makeBox(
        [-length/2, -width/2, 0],
        [length/2, width/2, height]
    );
    
    // Cutouts
    // Cutout at (unitSize * -9, 0)
    let cutout1 = makeBox(
        [-unitSize * 4, -unitSize * 3, 0],
        [unitSize * 4, unitSize * 3, unitHeight * 2]
    ).translate(unitSize * -9, 0, 0);
    part = part.cut(cutout1);
    
    // Cutout at (unitSize * 9, 0)
    let cutout2 = makeBox(
        [-unitSize * 4, -unitSize * 3, 0],
        [unitSize * 4, unitSize * 3, unitHeight * 2]
    ).translate(unitSize * 9, 0, 0);
    part = part.cut(cutout2);
    
    // Cutout at (0, 0) with fixed dimensions 44x22
    let cutout3 = makeBox(
        [-22, -11, 0],
        [22, 11, unitHeight * 2]
    ).translate(0, 0, 0);
    part = part.cut(cutout3);
    
    // Cutout at (0, unitSize * -2.5)
    let cutout4 = makeBox(
        [-unitSize * 13, -unitSize * 2.5, 0],
        [unitSize * 13, unitSize * 2.5, unitHeight * 2]
    ).translate(0, unitSize * -2.5, 0);
    part = part.cut(cutout4);
    
    // Cutout at (unitSize * -7.5, 0) at mid-height
    let cutout5 = makeBox(
        [-unitSize * 3.5, -unitSize * 5, 0],
        [unitSize * 3.5, unitSize * 5, unitHeight * 2]
    ).translate(unitSize * -7.5, 0, unitHeight);
    part = part.cut(cutout5);
    
    // Cutout at (unitSize * 7.5, 0) at mid-height
    let cutout6 = makeBox(
        [-unitSize * 3.5, -unitSize * 5, 0],
        [unitSize * 3.5, unitSize * 5, unitHeight * 2]
    ).translate(unitSize * 7.5, 0, unitHeight);
    part = part.cut(cutout6);
    
    // Cutout at (0, unitSize * 4) at mid-height
    let cutout7 = makeBox(
        [-unitSize * 4, -unitSize * 1, 0],
        [unitSize * 4, unitSize * 1, unitHeight * 2]
    ).translate(0, unitSize * 4, unitHeight);
    part = part.cut(cutout7);
    
    // Mounting screw holes (through holes)
    const screwRadius = 1.5;
    const screwHeight = unitHeight * 2;
    
    const screwPositions = [
        [unitSize * -12, unitSize * 4,  0],
        [unitSize * -12, unitSize * -4, 0],
        [unitSize * 12,  unitSize * 4,  0],
        [unitSize * 12,  unitSize * -4, 0],
    ];
    
    for (const pos of screwPositions) {
        let screwHole = makeCylinder(screwRadius, screwHeight, [0, 0, 0], [0, 0, screwHeight])
            .translate(pos[0], pos[1], pos[2]);
        part = part.cut(screwHole);
    }
    
    // Counterbore holes (recessed areas for screw heads)
    const counterboreRadius = 3;
    const counterboreHeight = unitHeight;
    const counterboreZ = unitHeight;
    
    const counterborePositions = [
        [unitSize * -12, unitSize * 4,  counterboreZ],
        [unitSize * -12, unitSize * -4, counterboreZ],
        [unitSize * 12,  unitSize * 4,  counterboreZ],
        [unitSize * 12,  unitSize * -4, counterboreZ],
    ];
    
    for (const pos of counterborePositions) {
        let counterbore = makeCylinder(counterboreRadius, counterboreHeight, [0, 0, 0], [0, 0, counterboreHeight])
            .translate(pos[0], pos[1], pos[2]);
        part = part.cut(counterbore);
    }
    
    // Apply fillet to all edges that are parallel to Z-axis
    part = part.fillet(2, (e) => e.inDirection('Z'));
    
    return part;
}